task_processing/plugins/mesos/execution_framework.py

```diff
@@ -170,29 +173,47 @@ def _background_check(self):
                         continue
 
                     if md.task_state == 'UNKNOWN':
-                        log.warning('Task {id} has been in unknown state for '
-                                    'longer than {timeout}. Re-enqueuing it.'
-                                    .format(
-                                        id=task_id,
-                                        timeout=self.task_staging_timeout_s
-                                    ))
+                        log.warning(
+                            f'Re-enqueuing task {task_id} in unknown state for '
+                            f'longer than {self.task_staging_timeout_s}'
+                        )
                         # Re-enqueue task
                         self.enqueue_task(md.task_config)
                         get_metric(
                             metrics.TASK_FAILED_TO_LAUNCH_COUNT).count(1)
                         continue
 
                     if md.task_state == 'TASK_STAGING':
-                        log.warning(
-                            'Killing stuck task {id}'.format(id=task_id)
-                        )
+                        log.warning(f'Killing stuck task {task_id}')
                         self.kill_task(task_id)
+                        self.task_metadata = self.task_metadata.set(
+                            task_id,
+                            md.set(
+                                task_state='TASK_STUCK',
+                                task_state_history=md.task_state_history.set(
+                                    'TASK_STUCK', time.time()),
+                            )
+                        )
                         self.blacklist_slave(
                             agent_id=self.task_metadata[task_id].agent_id,
                             timeout=self.slave_blacklist_timeout_s,
                         )
                         get_metric(metrics.TASK_STUCK_COUNT).count(1)
 
+                    if md.task_state == 'TASK_STUCK':
+                        t = time.time()
+                        # account for time we spent in this loop
+                        time_delta = t - time_now + 10
```

Comment:
to clarify this logic:

`time_delta` - 10 seconds we slept since the last iteration of background check started + seconds we spent in current iteration
`time_stuck` - seconds since task was put in TASK_STUCK state
`hour_rolled` - seconds since `time_stuck` crossed another hour boundary

basically we want to send kill request in the first 0..10s interval of each hour. there are still some unlikely timings when we will miss retrying a kill, but i really don't want to maintain any more state in this framework.